File: backend/src/models/message.py
"""
Message model for SMS/WhatsApp reminders.

Implements T157 from tasks.md.
"""
from datetime import datetime
from typing import Optional
import enum

from sqlalchemy import (
    Column,
    Integer,
    String,
    Text,
    DateTime,
    Enum as SQLEnum,
    ForeignKey,
    Boolean,
)
from sqlalchemy.orm import relationship, Mapped, mapped_column
from sqlalchemy.dialects.postgresql import JSONB

from ..core.database import Base


class MessageChannel(str, enum.Enum):
    """Message delivery channel."""
    SMS = "sms"
    WHATSAPP = "whatsapp"
    EMAIL = "email"


class MessageStatus(str, enum.Enum):
    """Message delivery status."""
    PENDING = "pending"
    SCHEDULED = "scheduled"
    SENDING = "sending"
    SENT = "sent"
    DELIVERED = "delivered"
    FAILED = "failed"
    CANCELLED = "cancelled"


class MessageTemplate(str, enum.Enum):
    """Predefined message templates."""
    PAYMENT_REMINDER = "payment_reminder"
    PAYMENT_OVERDUE = "payment_overdue"
    PAYMENT_RECEIVED = "payment_received"
    LEASE_EXPIRING = "lease_expiring"
    MAINTENANCE_NOTICE = "maintenance_notice"
    CUSTOM = "custom"


class Message(Base):
    """Message model for SMS/WhatsApp/Email communications.
    
    Features:
    - Multiple delivery channels (SMS, WhatsApp, Email)
    - Template-based messaging
    - Scheduled delivery
    - Delivery tracking and status updates
    - Bulk messaging support
    - Personalization with tenant data
    """
    
    __tablename__ = "messages"
    
    # Primary Key
    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    
    # Foreign Keys
    tenant_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("tenants.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
        comment="Recipient tenant"
    )
    sent_by: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=False,
        index=True,
        comment="User who sent the message"
    )
    property_id: Mapped[Optional[int]] = mapped_column(
        Integer,
        ForeignKey("properties.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
        comment="Related property (optional)"
    )
    
    # Message Content
    template: Mapped[MessageTemplate] = mapped_column(
        SQLEnum(MessageTemplate, name="message_template"),
        nullable=False,
        comment="Template used for message"
    )
    subject: Mapped[Optional[str]] = mapped_column(
        String(255),
        nullable=True,
        comment="Message subject (for email)"
    )
    content: Mapped[str] = mapped_column(
        Text,
        nullable=False,
        comment="Message body content"
    )
    
    # Delivery Configuration
    channel: Mapped[MessageChannel] = mapped_column(
        SQLEnum(MessageChannel, name="message_channel"),
        nullable=False,
        comment="Delivery channel"
    )
    recipient_phone: Mapped[Optional[str]] = mapped_column(
        String(20),
        nullable=True,
        comment="Phone number for SMS/WhatsApp"
    )
    recipient_email: Mapped[Optional[str]] = mapped_column(
        String(255),
        nullable=True,
        comment="Email address for email channel"
    )
    
    # Status and Tracking
    status: Mapped[MessageStatus] = mapped_column(
        SQLEnum(MessageStatus, name="message_status"),
        nullable=False,
        default=MessageStatus.PENDING,
        comment="Current message status"
    )
    
    # Scheduling
    scheduled_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        comment="Scheduled send time"
    )
    sent_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        comment="Actual send time"
    )
    delivered_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        comment="Delivery confirmation time"
    )
    
    # Error Handling
    error_message: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
        comment="Error details if failed"
    )
    retry_count: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
        comment="Number of retry attempts"
    )
    max_retries: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=3,
        comment="Maximum retry attempts"
    )
    
    # External Provider Data
    provider_message_id: Mapped[Optional[str]] = mapped_column(
        String(255),
        nullable=True,
        index=True,
        comment="External provider's message ID (Twilio SID, etc.)"
    )
    provider_response: Mapped[Optional[dict]] = mapped_column(
        JSONB,
        nullable=True,
        comment="Full provider response"
    )
    
    # Metadata
    metadata: Mapped[Optional[dict]] = mapped_column(
        JSONB,
        nullable=True,
        comment="Additional message metadata"
    )
    
    # Bulk Message Grouping
    bulk_message_id: Mapped[Optional[str]] = mapped_column(
        String(255),
        nullable=True,
        index=True,
        comment="ID for bulk message batch"
    )
    
    # Timestamps
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=datetime.utcnow,
        nullable=False,
        comment="Record creation timestamp"
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
        nullable=False,
        comment="Last update timestamp"
    )
    
    # Relationships
    tenant = relationship(
        "Tenant",
        back_populates="messages",
        foreign_keys=[tenant_id]
    )
    sender = relationship(
        "User",
        back_populates="sent_messages",
        foreign_keys=[sent_by]
    )
    property = relationship(
        "Property",
        back_populates="messages",
        foreign_keys=[property_id]
    )
    
    def __repr__(self) -> str:
        return f"<Message(id={self.id}, tenant_id={self.tenant_id}, channel={self.channel}, status={self.status})>"
    
    @property
    def is_pending(self) -> bool:
        """Check if message is pending delivery."""
        return self.status in (MessageStatus.PENDING, MessageStatus.SCHEDULED)
    
    @property
    def is_sent(self) -> bool:
        """Check if message has been sent."""
        return self.status in (MessageStatus.SENT, MessageStatus.DELIVERED)
    
    @property
    def has_failed(self) -> bool:
        """Check if message delivery failed."""
        return self.status == MessageStatus.FAILED
    
    @property
    def can_retry(self) -> bool:
        """Check if message can be retried."""
        return self.has_failed and self.retry_count < self.max_retries

File: mobile/lib/models/message.dart
/// Message model for SMS/WhatsApp/Email communications.
///
/// Implements T168 from tasks.md.
library;

enum MessageChannel {
  sms,
  whatsapp,
  email,
}

enum MessageStatus {
  pending,
  scheduled,
  sending,
  sent,
  delivered,
  failed,
  cancelled,
}

enum MessageTemplate {
  paymentReminder,
  paymentOverdue,
  paymentReceived,
  leaseExpiring,
  maintenanceNotice,
  custom,
}

class Message {
  final int id;
  final int tenantId;
  final int? sentBy;
  final int? propertyId;
  final MessageTemplate template;
  final String? subject;
  final String content;
  final MessageChannel channel;
  final String? recipientPhone;
  final String? recipientEmail;
  final MessageStatus status;
  final DateTime? scheduledAt;
  final DateTime? sentAt;
  final DateTime? deliveredAt;
  final String? errorMessage;
  final int retryCount;
  final String? providerMessageId;
  final Map<String, dynamic>? providerResponse;
  final String? bulkMessageId;
  final Map<String, dynamic>? metadata;
  final DateTime createdAt;
  final DateTime? updatedAt;

  // Computed properties
  bool get isPending => status == MessageStatus.pending;
  bool get isScheduled => status == MessageStatus.scheduled;
  bool get isSent => status == MessageStatus.sent;
  bool get isDelivered => status == MessageStatus.delivered;
  bool get hasFailed => status == MessageStatus.failed;
  bool get canRetry => hasFailed && retryCount < 3;

  Message({
    required this.id,
    required this.tenantId,
    this.sentBy,
    this.propertyId,
    required this.template,
    this.subject,
    required this.content,
    required this.channel,
    this.recipientPhone,
    this.recipientEmail,
    required this.status,
    this.scheduledAt,
    this.sentAt,
    this.deliveredAt,
    this.errorMessage,
    required this.retryCount,
    this.providerMessageId,
    this.providerResponse,
    this.bulkMessageId,
    this.metadata,
    required this.createdAt,
    this.updatedAt,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['id'] as int,
      tenantId: json['tenant_id'] as int,
      sentBy: json['sent_by'] as int?,
      propertyId: json['property_id'] as int?,
      template: _templateFromString(json['template'] as String),
      subject: json['subject'] as String?,
      content: json['content'] as String,
      channel: _channelFromString(json['channel'] as String),
      recipientPhone: json['recipient_phone'] as String?,
      recipientEmail: json['recipient_email'] as String?,
      status: _statusFromString(json['status'] as String),
      scheduledAt: json['scheduled_at'] != null
          ? DateTime.parse(json['scheduled_at'] as String)
          : null,
      sentAt: json['sent_at'] != null
          ? DateTime.parse(json['sent_at'] as String)
          : null,
      deliveredAt: json['delivered_at'] != null
          ? DateTime.parse(json['delivered_at'] as String)
          : null,
      errorMessage: json['error_message'] as String?,
      retryCount: json['retry_count'] as int? ?? 0,
      providerMessageId: json['provider_message_id'] as String?,
      providerResponse: json['provider_response'] as Map<String, dynamic>?,
      bulkMessageId: json['bulk_message_id'] as String?,
      metadata: json['metadata'] as Map<String, dynamic>?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'tenant_id': tenantId,
      'sent_by': sentBy,
      'property_id': propertyId,
      'template': _templateToString(template),
      'subject': subject,
      'content': content,
      'channel': _channelToString(channel),
      'recipient_phone': recipientPhone,
      'recipient_email': recipientEmail,
      'status': _statusToString(status),
      'scheduled_at': scheduledAt?.toIso8601String(),
      'sent_at': sentAt?.toIso8601String(),
      'delivered_at': deliveredAt?.toIso8601String(),
      'error_message': errorMessage,
      'retry_count': retryCount,
      'provider_message_id': providerMessageId,
      'provider_response': providerResponse,
      'bulk_message_id': bulkMessageId,
      'metadata': metadata,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }

  Message copyWith({
    int? id,
    int? tenantId,
    int? sentBy,
    int? propertyId,
    MessageTemplate? template,
    String? subject,
    String? content,
    MessageChannel? channel,
    String? recipientPhone,
    String? recipientEmail,
    MessageStatus? status,
    DateTime? scheduledAt,
    DateTime? sentAt,
    DateTime? deliveredAt,
    String? errorMessage,
    int? retryCount,
    String? providerMessageId,
    Map<String, dynamic>? providerResponse,
    String? bulkMessageId,
    Map<String, dynamic>? metadata,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Message(
      id: id ?? this.id,
      tenantId: tenantId ?? this.tenantId,
      sentBy: sentBy ?? this.sentBy,
      propertyId: propertyId ?? this.propertyId,
      template: template ?? this.template,
      subject: subject ?? this.subject,
      content: content ?? this.content,
      channel: channel ?? this.channel,
      recipientPhone: recipientPhone ?? this.recipientPhone,
      recipientEmail: recipientEmail ?? this.recipientEmail,
      status: status ?? this.status,
      scheduledAt: scheduledAt ?? this.scheduledAt,
      sentAt: sentAt ?? this.sentAt,
      deliveredAt: deliveredAt ?? this.deliveredAt,
      errorMessage: errorMessage ?? this.errorMessage,
      retryCount: retryCount ?? this.retryCount,
      providerMessageId: providerMessageId ?? this.providerMessageId,
      providerResponse: providerResponse ?? this.providerResponse,
      bulkMessageId: bulkMessageId ?? this.bulkMessageId,
      metadata: metadata ?? this.metadata,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  // Helper methods for enum conversion
  static MessageChannel _channelFromString(String value) {
    switch (value) {
      case 'sms':
        return MessageChannel.sms;
      case 'whatsapp':
        return MessageChannel.whatsapp;
      case 'email':
        return MessageChannel.email;
      default:
        throw ArgumentError('Invalid channel: $value');
    }
  }

  static String _channelToString(MessageChannel channel) {
    switch (channel) {
      case MessageChannel.sms:
        return 'sms';
      case MessageChannel.whatsapp:
        return 'whatsapp';
      case MessageChannel.email:
        return 'email';
    }
  }

  static MessageStatus _statusFromString(String value) {
    switch (value) {
      case 'pending':
        return MessageStatus.pending;
      case 'scheduled':
        return MessageStatus.scheduled;
      case 'sending':
        return MessageStatus.sending;
      case 'sent':
        return MessageStatus.sent;
      case 'delivered':
        return MessageStatus.delivered;
      case 'failed':
        return MessageStatus.failed;
      case 'cancelled':
        return MessageStatus.cancelled;
      default:
        throw ArgumentError('Invalid status: $value');
    }
  }

  static String _statusToString(MessageStatus status) {
    switch (status) {
      case MessageStatus.pending:
        return 'pending';
      case MessageStatus.scheduled:
        return 'scheduled';
      case MessageStatus.sending:
        return 'sending';
      case MessageStatus.sent:
        return 'sent';
      case MessageStatus.delivered:
        return 'delivered';
      case MessageStatus.failed:
        return 'failed';
      case MessageStatus.cancelled:
        return 'cancelled';
    }
  }

  static MessageTemplate _templateFromString(String value) {
    switch (value) {
      case 'payment_reminder':
        return MessageTemplate.paymentReminder;
      case 'payment_overdue':
        return MessageTemplate.paymentOverdue;
      case 'payment_received':
        return MessageTemplate.paymentReceived;
      case 'lease_expiring':
        return MessageTemplate.leaseExpiring;
      case 'maintenance_notice':
        return MessageTemplate.maintenanceNotice;
      case 'custom':
        return MessageTemplate.custom;
      default:
        throw ArgumentError('Invalid template: $value');
    }
  }

  static String _templateToString(MessageTemplate template) {
    switch (template) {
      case MessageTemplate.paymentReminder:
        return 'payment_reminder';
      case MessageTemplate.paymentOverdue:
        return 'payment_overdue';
      case MessageTemplate.paymentReceived:
        return 'payment_received';
      case MessageTemplate.leaseExpiring:
        return 'lease_expiring';
      case MessageTemplate.maintenanceNotice:
        return 'maintenance_notice';
      case MessageTemplate.custom:
        return 'custom';
    }
  }
}

/// Statistics for message delivery
class MessageStatistics {
  final int totalMessages;
  final int sent;
  final int delivered;
  final int failed;
  final int pending;
  final double deliveryRate;

  MessageStatistics({
    required this.totalMessages,
    required this.sent,
    required this.delivered,
    required this.failed,
    required this.pending,
    required this.deliveryRate,
  });

  factory MessageStatistics.fromJson(Map<String, dynamic> json) {
    return MessageStatistics(
      totalMessages: json['total_messages'] as int,
      sent: json['sent'] as int,
      delivered: json['delivered'] as int,
      failed: json['failed'] as int,
      pending: json['pending'] as int,
      deliveryRate: (json['delivery_rate'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'total_messages': totalMessages,
      'sent': sent,
      'delivered': delivered,
      'failed': failed,
      'pending': pending,
      'delivery_rate': deliveryRate,
    };
  }
}
